internal/characters/thoma/abil.go

```diff
@@ -0,0 +1,237 @@
+		ICDGroup:   core.ICDGroupDefault,
+		StrikeType: core.StrikeTypeDefault,
+		Element:    core.Pyro,
+		Durability: 50,
+		Mult:       burst[c.TalentLvlSkill()],
+	}
+
+	// damage component not final
+	x, y := c.Core.Targets[0].Shape().Pos()
+	c.Core.Combat.QueueAttack(ai, core.NewCircleHit(x, y, 2, false, core.TargettableEnemy), f+1, f+1)
+	
+	d := 15
+	if c.Base.Cons >= 2 {
+		d = 18
+	}
+
+	c.Core.Status.AddStatus("thomaburst", d*60)
+
+	c.burstProc()
+	
+	if c.Base.Cons >= 4 {
+		c.AddTask(func() {
+			c.c4Restore()
+		}, "thoma-c4-restore", 15)
+	}
+
+	cd := 20
+	if c.Base.Cons >= 1 {
+		cd = 17; //the CD reduction activates when a character protected by Thoma's shield is hit. Since it is almost impossible for this not to activate, we set the duration to 17 for sim purposes.
+	}
+	c.SetCDWithDelay(core.ActionBurst, cd*60, 11)
+	c.ConsumeEnergy(11)
+
+	return f, a
+}
+
+func (c *char) burstProc() {
+	// does not deactivate on death
+	icd := 0
+	c.Core.Events.Subscribe(core.OnDamage, func(args ...interface{}) bool {
+		ae := args[1].(*core.AttackEvent)
+		t := args[0].(core.Target)
+		if ae.Info.AttackTag != core.AttackTagNormal && ae.Info.AttackTag != core.AttackTagExtra {
+			return false
+		}
+		if c.Core.Status.Duration("thomaburst") == 0 {
+			return false
+		}
+		if icd > c.Core.F {
+			c.Core.Log.NewEvent("thoma Q (active) on icd", core.LogCharacterEvent, c.Index, "frame", c.Core.F)
+			return false
+		}
+
+		ai := core.AttackInfo{
+			ActorIndex: c.Index,
+			Abil:       "Fiery Collapse",
+			AttackTag:  core.AttackTagElementalBurst,
+			ICDTag:     core.ICDTagElementalBurst,
+			ICDGroup:   core.ICDGroupDefault,
+			StrikeType: core.StrikeTypeDefault,
+			Element:    core.Pyro,
+			Durability: 50,
+			Mult:       burstproc[c.TalentLvlSkill()],
+			FlatDmg:    0.022 * c.HPMax,
+		}
+		//trigger a chain of attacks starting at the first target
+		atk := core.AttackEvent{
+			Info: ai,
+		}
+		atk.SourceFrame = c.Core.F
+		atk.Pattern = core.NewDefSingleTarget(t.Index(), core.TargettableEnemy)
+		cb := func(a core.AttackCB) {
+			shieldamt := (burstshieldpp[c.TalentLvlSkill()]*c.MaxHP() + burstshieldflat[c.TalentLvlSkill()])
+			c.genShield("Thoma Burst", shieldamt)
+		}
+		if cb != nil {
+			atk.Callbacks = append(atk.Callbacks, cb)
+		}
+		c.Core.Combat.QueueAttackEvent(&atk, 1)
+
+		c.Core.Log.NewEvent("thoma Q proc'd", core.LogCharacterEvent, c.Index, "frame", c.Core.F, "char", ae.Info.ActorIndex, "attack tag", ae.Info.AttackTag)
+
+		icd = c.Core.F + 60 // once per second
+		return false
+	}, "thoma-burst")
+}
+
+func (c *char) genShield(src string, shieldamt float64) {
+	if c.Core.F > c.a1icd && c.a1Stack < 5 {
+		c.a1Stack++
+		c.a1icd = c.Core.F + 0.3*60
+		c.Core.Status.AddStatus("thoma-a1", 6*60)
+	}
+	if c.Core.Shields.Get(core.ShieldThomaSkill) != nil {
+		if c.Core.Shields.Get(core.ShieldThomaSkill).CurrentHP()+shieldamt > c.MaxShield {
+			shieldamt = c.MaxShield - c.Core.Shields.Get(core.ShieldThomaSkill).CurrentHP()
```

Comment:
actually I think I made a fundamental misunderstanding in how shields are generated in gcsim so this is quite wrong. will have to look into this more but feel free to PR if you already have looked into it

Reply:
Upon closer inspection, it seems like the way the Add method works is that it adds an entirely new shield (or replaces it) so will need to figure out how to stack the shield instead of replacing it.

Reply:
I think that we could do something like:

```
if c.Core.Shields.Get(core.ShieldThomaSkill).CurrentHP()+shieldamt > c.MaxShield {
			shieldamt = c.MaxShield - c.Core.Shields.Get(core.ShieldThomaSkill).CurrentHP()
} else {
                        shieldamt += c.Core.Shields.Get(core.ShieldThomaSkill).CurrentHP()
}
```

I haven't tested it yet though, if I find some time after work I can try later today